### ISSUE TITLE
Build error: invalid access to non-static data member 'sdch::Config::enable'  of NULL object

### ISSUE THREAD

**Issue by redburn** [bacek/sdch_module#2]
I'm trying to build the sdch module into nginx but keep running into errors:

> cc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic  -I src/core -I src/event -I src/event/modules -I src/os/unix -I src/http/modules/perl -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I /usr/include/libxml2 -I objs -I src/http -I src/http/modules -I src/http/v2 -I src/mail -I src/stream \
>                 -o objs/addon/sdch_module/sdch_config.o \
>                 /tmp/sdch_module/sdch_config.cc
> cc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic  -I src/core -I src/event -I src/event/modules -I src/os/unix -I src/http/modules/perl -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I /usr/include/libxml2 -I objs -I src/http -I src/http/modules -I src/http/v2 -I src/mail -I src/stream \
>                 -o objs/addon/sdch_module/sdch_autoauto_handler.o \
>                 /tmp/sdch_module/sdch_autoauto_handler.cc
> cc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic  -I src/core -I src/event -I src/event/modules -I src/os/unix -I src/http/modules/perl -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I /usr/include/libxml2 -I objs -I src/http -I src/http/modules -I src/http/v2 -I src/mail -I src/stream \
>                 -o objs/addon/sdch_module/sdch_dictionary.o \
>                 /tmp/sdch_module/sdch_dictionary.cc
> cc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic  -I src/core -I src/event -I src/event/modules -I src/os/unix -I src/http/modules/perl -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I /usr/include/libxml2 -I objs -I src/http -I src/http/modules -I src/http/v2 -I src/mail -I src/stream \
>                 -o objs/addon/sdch_module/sdch_dictionary_factory.o \
>                 /tmp/sdch_module/sdch_dictionary_factory.cc
> cc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic  -I src/core -I src/event -I src/event/modules -I src/os/unix -I src/http/modules/perl -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I /usr/include/libxml2 -I objs -I src/http -I src/http/modules -I src/http/v2 -I src/mail -I src/stream \
>                 -o objs/addon/sdch_module/sdch_dump_handler.o \
>                 /tmp/sdch_module/sdch_dump_handler.cc
> cc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic  -I src/core -I src/event -I src/event/modules -I src/os/unix -I src/http/modules/perl -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I /usr/include/libxml2 -I objs -I src/http -I src/http/modules -I src/http/v2 -I src/mail -I src/stream \
>                 -o objs/addon/sdch_module/sdch_encoding_handler.o \
>                 /tmp/sdch_module/sdch_encoding_handler.cc
> cc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic  -I src/core -I src/event -I src/event/modules -I src/os/unix -I src/http/modules/perl -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I /usr/include/libxml2 -I objs -I src/http -I src/http/modules -I src/http/v2 -I src/mail -I src/stream \
>                 -o objs/addon/sdch_module/sdch_fastdict_factory.o \
>                 /tmp/sdch_module/sdch_fastdict_factory.cc
> cc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic  -I src/core -I src/event -I src/event/modules -I src/os/unix -I src/http/modules/perl -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I /usr/include/libxml2 -I objs -I src/http -I src/http/modules -I src/http/v2 -I src/mail -I src/stream \
>                 -o objs/addon/sdch_module/sdch_handler.o \
>                 /tmp/sdch_module/sdch_handler.cc
> cc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic  -I src/core -I src/event -I src/event/modules -I src/os/unix -I src/http/modules/perl -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I /usr/include/libxml2 -I objs -I src/http -I src/http/modules -I src/http/v2 -I src/mail -I src/stream \
>                 -o objs/addon/sdch_module/sdch_main_config.o \
>                 /tmp/sdch_module/sdch_main_config.cc
> cc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic  -I src/core -I src/event -I src/event/modules -I src/os/unix -I src/http/modules/perl -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I /usr/include/libxml2 -I objs -I src/http -I src/http/modules -I src/http/v2 -I src/mail -I src/stream \
>                 -o objs/addon/sdch_module/sdch_module.o \
>                 /tmp/sdch_module/sdch_module.cc
> cc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic  -I src/core -I src/event -I src/event/modules -I src/os/unix -I src/http/modules/perl -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I njs-0.1.0/nginx/../nxt -I njs-0.1.0/nginx/../njs -I /usr/include/libxml2 -I objs -I src/http -I src/http/modules -I src/http/v2 -I src/mail -I src/stream \
>                 -o objs/addon/sdch_module/sdch_output_handler.o \
>                 /tmp/sdch_module/sdch_output_handler.cc
> cc1plus: warnings being treated as errors
> /tmp/sdch_module/sdch_module.cc:85: error: invalid access to non-static data member 'sdch::Config::enable'  of NULL object
> /tmp/sdch_module/sdch_module.cc:85: error: (perhaps the 'offsetof' macro was used incorrectly)
> /tmp/sdch_module/sdch_module.cc:94: error: invalid access to non-static data member 'sdch::Config::sdch_disablecv_s'  of NULL object
> /tmp/sdch_module/sdch_module.cc:94: error: (perhaps the 'offsetof' macro was used incorrectly)
> /tmp/sdch_module/sdch_module.cc:101: error: invalid access to non-static data member 'sdch::Config::bufs'  of NULL object
> /tmp/sdch_module/sdch_module.cc:101: error: (perhaps the 'offsetof' macro was used incorrectly)
> /tmp/sdch_module/sdch_module.cc:119: error: invalid access to non-static data member 'sdch::Config::sdch_group'  of NULL object
> /tmp/sdch_module/sdch_module.cc:119: error: (perhaps the 'offsetof' macro was used incorrectly)
> /tmp/sdch_module/sdch_module.cc:128: error: invalid access to non-static data member 'sdch::Config::sdch_url'  of NULL object
> /tmp/sdch_module/sdch_module.cc:128: error: (perhaps the 'offsetof' macro was used incorrectly)
> /tmp/sdch_module/sdch_module.cc:137: error: invalid access to non-static data member 'sdch::Config::sdch_dumpdir'  of NULL object
> /tmp/sdch_module/sdch_module.cc:137: error: (perhaps the 'offsetof' macro was used incorrectly)
> /tmp/sdch_module/sdch_module.cc:146: error: invalid access to non-static data member 'sdch::Config::sdch_proxied'  of NULL object
> /tmp/sdch_module/sdch_module.cc:146: error: (perhaps the 'offsetof' macro was used incorrectly)
> /tmp/sdch_module/sdch_module.cc:156: error: invalid access to non-static data member 'sdch::Config::types_keys'  of NULL object
> /tmp/sdch_module/sdch_module.cc:156: error: (perhaps the 'offsetof' macro was used incorrectly)
> /tmp/sdch_module/sdch_module.cc:165: error: invalid access to non-static data member 'sdch::Config::nodict_types_keys'  of NULL object
> /tmp/sdch_module/sdch_module.cc:165: error: (perhaps the 'offsetof' macro was used incorrectly)
> /tmp/sdch_module/sdch_module.cc:174: error: invalid access to non-static data member 'sdch::Config::enable_fastdict'  of NULL object
> /tmp/sdch_module/sdch_module.cc:174: error: (perhaps the 'offsetof' macro was used incorrectly)
> /tmp/sdch_module/sdch_module.cc:183: error: invalid access to non-static data member 'sdch::Config::min_length'  of NULL object
> /tmp/sdch_module/sdch_module.cc:183: error: (perhaps the 'offsetof' macro was used incorrectly)
> /tmp/sdch_module/sdch_module.cc:191: error: invalid access to non-static data member 'sdch::Config::vary'  of NULL object
> /tmp/sdch_module/sdch_module.cc:191: error: (perhaps the 'offsetof' macro was used incorrectly)
> /tmp/sdch_module/sdch_module.cc:198: error: invalid access to non-static data member 'sdch::MainConfig::stor_size'  of NULL object
> /tmp/sdch_module/sdch_module.cc:198: error: (perhaps the 'offsetof' macro was used incorrectly)
> make[1]: **\* [objs/addon/sdch_module/sdch_module.o] Error 1
> make[1]: **\* Waiting for unfinished jobs....
> make[1]: Leaving directory `/root/rpmbuild/BUILD/nginx-1.11.3'
> make: **\* [build] Error 2
> error: Bad exit status from /var/tmp/rpm-tmp.eMzQ31 (%build)
> 
> RPM build errors:
>     Bad exit status from /var/tmp/rpm-tmp.eMzQ31 (%build)

Any idea what the problem could be?

**Comment by wawaAtYandex** [comment]
I can not reproduce the issue. Could you please let me know the nginx version?

**Comment by redburn** [comment]
nginx/1.11.3 (mainline)

Building on CentOS 6.8 x64 with open-vcdiff 0.8.3.

gcc version 4.4.7 20120313 (Red Hat 4.4.7-17) (GCC)

**Comment by redburn** [comment]
Same with nginx/1.10.1 (stable) and nginx/1.8.0 (stable).

Does seem to work on CentOS 7...

**Comment by wawaAtYandex** [comment]
Let us reduce extra stuff.
Pls, try smth like this:
cc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g   -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules \
        -o objs/addon/sdch_module/sdch_output_handler.o \
        tmp/sdch_module/sdch_output_handler.cc

**Comment by redburn** [comment]
I got it to build by suppressing the warnings regarding offsetof (added -Wno-invalid-offsetof to build flags) instead of treating them like errors (removed -Werror). This isn't ideal (I can't easily use rpmbuild, for example), but it's fine for now while I'm experimenting with sdch.

CentOS 7 has a newer version of glibc (2.17) than CentOS 6 (2.12). I'm not a C/C++ programmer, but from my readings I would suspect that something may have changed in how strict the compiler is about using offsetof on non-POD structures. From the documentation on -Wno-invalid-offsetof:

> Suppress warnings from applying the 'offsetof' macro to a non-POD type. According to the 1998 ISO C++ standard, applying 'offsetof' to a non-POD type is undefined. In existing C++ implementations, however, 'offsetof' typically gives meaningful results even when applied to certain kinds of non-POD types. (Such as a simple 'struct' that fails to be a POD type only by virtue of having a constructor.) This flag is for users who are aware that they are writing nonportable code and who have deliberately chosen to ignore the warning about it.
> 
> The restrictions on 'offsetof' may be relaxed in a future version of the C++ standard. 

If there's an alternative to offsetof for use in sdch_module.cc, that would probably solve this too.

**Comment by redburn** [comment]
I cheered too soon, nginx segfaults when I enable the sdch module in the configuration files, specifically when referring to the sdch_dict file. Same on CentOS 7, where it builds just fine.

**Comment by bacek** [comment]
Hello.

Can you try to run nginx with gdb and provide backtrace when it's crashed?

Something like

$ gdb --args /path/to/nginx -c /path/to/nginx.conf

and run "bt" when it's crashed.

(I don't have CentOS installation right now and it will take some time to install.)

**Comment by bacek** [comment]
Actually I managed to compile it on CentOS 6.8 (with same workaround). Can you provide you nginx config because I can't crash it easily with our sample one?

**Comment by redburn** [comment]
```
(gdb) run
Starting program: /usr/sbin/nginx -c /etc/nginx/nginx.conf
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
get_dict_payload (dictend=0xa57140 "",
    dictbegin=0xa47140 "?page=4\">4</a></li><li><a href=\"/screensand-1.png\" width=\"207\" height=\"155\" alt=\"Screenshot of s-1.png\" width=\"207\" height=\"155\" alt=\"Screenshot of Wx</span> <span class=\"info\">Free, for Windows</span"...) at /tmp/sdch_module/sdch_dictionary_factory.cc:27
27                      if (*nl == '\n')
```

```
(gdb) bt
#0  get_dict_payload (dictend=0xa57140 "",
    dictbegin=0xa47140 "?page=4\">4</a></li><li><a href=\"/screensand-1.png\" width=\"207\" height=\"155\" alt=\"Screenshot of s-1.png\" width=\"207\" height=\"155\" alt=\"Screenshot of Wx</span> <span class=\"info\">Free, for Windows</span"...) at /tmp/sdch_module/sdch_dictionary_factory.cc:27
#1  sdch::DictionaryFactory::load_dictionary (this=0xa441e0, filename=0xa451e9 "/home/example.com/public_html/sdch_dict") at /tmp/sdch_module/sdch_dictionary_factory.cc:85
#2  0x000000000050156c in sdch::set_sdch_dict (cf=0x7fffffffe040, cmd=<optimized out>, cnf=0xa440a0) at /tmp/sdch_module/sdch_module.cc:875
#3  0x0000000000465ec3 in ngx_conf_handler (last=0, cf=0x7fffffffe040) at src/core/ngx_conf_file.c:427
#4  ngx_conf_parse (cf=cf@entry=0x7fffffffe040, filename=filename@entry=0x0) at src/core/ngx_conf_file.c:283
#5  0x000000000048865d in ngx_http_core_location (cf=0x7fffffffe040, cmd=<optimized out>, dummy=<optimized out>) at src/http/ngx_http_core_module.c:3256
#6  0x0000000000465ec3 in ngx_conf_handler (last=1, cf=0x7fffffffe040) at src/core/ngx_conf_file.c:427
#7  ngx_conf_parse (cf=cf@entry=0x7fffffffe040, filename=filename@entry=0x0) at src/core/ngx_conf_file.c:283
#8  0x000000000048745b in ngx_http_core_server (cf=0x7fffffffe040, cmd=<optimized out>, dummy=<optimized out>) at src/http/ngx_http_core_module.c:3028
#9  0x0000000000465ec3 in ngx_conf_handler (last=1, cf=0x7fffffffe040) at src/core/ngx_conf_file.c:427
#10 ngx_conf_parse (cf=cf@entry=0x7fffffffe040, filename=filename@entry=0x7fffffffd950) at src/core/ngx_conf_file.c:283
#11 0x00000000004661f5 in ngx_conf_include (cf=0x7fffffffe040, cmd=<optimized out>, conf=<optimized out>) at src/core/ngx_conf_file.c:831
#12 0x0000000000465ec3 in ngx_conf_handler (last=0, cf=0x7fffffffe040) at src/core/ngx_conf_file.c:427
#13 ngx_conf_parse (cf=cf@entry=0x7fffffffe040, filename=filename@entry=0x0) at src/core/ngx_conf_file.c:283
#14 0x00000000004823ba in ngx_http_block (cf=0x7fffffffe040, cmd=<optimized out>, conf=0x9fa368) at src/http/ngx_http.c:237
#15 0x0000000000465ec3 in ngx_conf_handler (last=1, cf=0x7fffffffe040) at src/core/ngx_conf_file.c:427
#16 ngx_conf_parse (cf=cf@entry=0x7fffffffe040, filename=filename@entry=0x9f9558) at src/core/ngx_conf_file.c:283
#17 0x0000000000463611 in ngx_init_cycle (old_cycle=old_cycle@entry=0x7fffffffe1f0) at src/core/ngx_cycle.c:274
#18 0x0000000000451ad4 in main (argc=<optimized out>, argv=<optimized out>) at src/core/nginx.c:276
```

/etc/nginx/nginx.conf

```
user  nginx;
worker_processes  auto;
worker_rlimit_nofile 65536;

error_log  /var/log/nginx/error.log warn;
pid        /var/run/nginx.pid;

events {
    worker_connections  2048;
}

http {
    include       /etc/nginx/mime.types;
    default_type  application/octet-stream;

    log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" "$http_x_forwarded_for"';

    access_log  /var/log/nginx/access.log  main;

    sendfile        on;
    #tcp_nopush     on;

    include /etc/nginx/conf.d/*.conf;

    keepalive_timeout   30;

    fastcgi_buffers     8 16k;
    fastcgi_buffer_size 32k;

    gzip                on;
    gzip_types          text/plain text/css application/x-javascript text/xml application/xml application/xml+rss text/javascript application/javascript application/vnd.ms-fontobject application/x-font-ttf;
    gzip_comp_level     5;
    gzip_vary           on;
    gzip_proxied        any;
    gzip_http_version   1.0;

    server_tokens       off;
    etag                off;

}
```

/etc/nginx/conf.d/example.conf

```
server {

        listen                  80;

        server_name             www.example.com example.com;

        access_log              /home/example.com/logs/access.log;
        error_log               /home/example.com/logs/error.log;

        root                    /home/example.com/public_html;

        index                   index.php;
        charset                 utf-8;

        location ~* \.php$ {

                try_files       $uri =404;
                fastcgi_pass    127.0.0.1:9000;
                fastcgi_param   SCRIPT_FILENAME $document_root$fastcgi_script_name;
                fastcgi_param   PATH_INFO $fastcgi_script_name;
                fastcgi_read_timeout    600s;
                include         /etc/nginx/fastcgi_params;

                sdch            on;
                sdch_dict       /home/example.com/public_html/sdch_dict;
                sdch_url        http://www.example.com/sdch_dict;

        }

}
```

> nginx version: nginx/1.11.3
> built by gcc 4.8.5 20150623 (Red Hat 4.8.5-4) (GCC)
> built with OpenSSL 1.0.2h  3 May 2016
> TLS SNI support enabled
> configure arguments: --prefix=/etc/nginx --sbin-path=/usr/sbin/nginx --modules-path=/usr/lib64/nginx/modules --conf-path=/etc/nginx/nginx.conf --error-log-path=/var/log/nginx/error.log --http-log-path=/var/log/nginx/access.log --pid-path=/var/run/nginx.pid --lock-path=/var/run/nginx.lock --http-client-body-temp-path=/var/cache/nginx/client_temp --http-proxy-temp-path=/var/cache/nginx/proxy_temp --http-fastcgi-temp-path=/var/cache/nginx/fastcgi_temp --http-uwsgi-temp-path=/var/cache/nginx/uwsgi_temp --http-scgi-temp-path=/var/cache/nginx/scgi_temp --user=nginx --group=nginx --with-http_ssl_module --with-http_v2_module --with-openssl=/tmp/openssl-1.0.2h --add-module=/tmp/ngx_brotli --add-module=/tmp/sdch_module --with-http_realip_module --with-http_addition_module --with-http_sub_module --with-http_dav_module --with-http_flv_module --with-http_mp4_module --with-http_gunzip_module --with-http_gzip_static_module --with-http_random_index_module --with-http_secure_link_module --with-http_stub_status_module --with-http_auth_request_module --with-http_xslt_module=dynamic --with-http_image_filter_module=dynamic --with-http_geoip_module=dynamic --with-http_perl_module=dynamic --add-dynamic-module=njs-0.1.0/nginx --with-threads --with-stream --with-stream_ssl_module --with-stream_geoip_module=dynamic --with-http_slice_module --with-mail --with-mail_ssl_module --with-file-aio --with-ipv6 --with-http_v2_module --with-cc-opt='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic'

**Comment by bacek** [comment]
Ok, crash is different issue (#3).

**Comment by bacek** [comment]
@redburn crash will be fixed in #3. You have mis-formed dictionary (without SDCH headers). If you update your dictionary to have something like 

```
Host: example.com

<content>
```

it should work.

**Comment by redburn** [comment]
Completely forgot about those headers. Thank you!

I assume "Host:" in your example should be "Domain:" instead.
